File: Languages/python/low-level-design-lld/design-patterns/proxy.py
"""
Proxy Pattern

Provides a surrogate or placeholder for another object to control access.
CachingProxy wraps a slow DatabaseQuery with lazy loading and caching.
"""

from abc import ABC, abstractmethod
import time


class DataSource(ABC):
    @abstractmethod
    def query(self, sql: str) -> list[dict]:
        pass


class SlowDatabaseQuery(DataSource):
    """Simulates a slow database with artificial delay."""

    def __init__(self):
        self._data = {
            "SELECT * FROM users": [
                {"id": 1, "name": "Alice"},
                {"id": 2, "name": "Bob"},
            ],
            "SELECT * FROM orders": [
                {"id": 1, "total": 99.99},
            ],
        }

    def query(self, sql: str) -> list[dict]:
        time.sleep(0.1)  # simulate latency
        return self._data.get(sql, [])


class CachingProxy(DataSource):
    """Caches query results to avoid repeated slow lookups."""

    def __init__(self, source: DataSource | None = None):
        self._source = source  # lazy: may be set later
        self._cache: dict[str, list[dict]] = {}
        self._hits = 0

    @property
    def source(self) -> DataSource:
        if self._source is None:
            self._source = SlowDatabaseQuery()  # lazy loading
        return self._source

    def query(self, sql: str) -> list[dict]:
        if sql in self._cache:
            self._hits += 1
            return self._cache[sql]
        result = self.source.query(sql)
        self._cache[sql] = result
        return result

    def invalidate(self, sql: str | None = None) -> None:
        if sql:
            self._cache.pop(sql, None)
        else:
            self._cache.clear()


if __name__ == "__main__":
    proxy = CachingProxy()  # lazy: no DB created yet
    assert proxy._source is None

    # First query: triggers lazy loading + actual query
    start = time.time()
    result = proxy.query("SELECT * FROM users")
    first_duration = time.time() - start
    assert len(result) == 2
    assert proxy._source is not None

    # Second query: served from cache (fast)
    start = time.time()
    cached = proxy.query("SELECT * FROM users")
    cache_duration = time.time() - start
    assert cached == result
    assert cache_duration < first_duration
    assert proxy._hits == 1

    # Invalidate and re-query
    proxy.invalidate("SELECT * FROM users")
    proxy.query("SELECT * FROM users")
    assert proxy._hits == 1  # cache miss after invalidation

    print("All tests passed!")

File: Languages/go/low-level-design-lld/design-patterns/proxy.go
package main

import (
	"fmt"
	"time"
)

// Proxy Pattern: Provides a surrogate or placeholder for another object
// to control access, add caching, logging, or lazy initialization.

// Subject interface
type Database interface {
	Query(sql string) string
	Execute(sql string) bool
}

// RealSubject - actual database
type RealDatabase struct {
	name string
}

func NewRealDatabase(name string) *RealDatabase {
	fmt.Printf("[RealDB] Connecting to %s...\n", name)
	return &RealDatabase{name: name}
}

func (db *RealDatabase) Query(sql string) string {
	return fmt.Sprintf("Result from %s: [rows for '%s']", db.name, sql)
}

func (db *RealDatabase) Execute(sql string) bool {
	fmt.Printf("[RealDB] Executing: %s\n", sql)
	return true
}

// CachingProxy - adds caching layer on top
type CachingProxy struct {
	db    *RealDatabase
	cache map[string]cachedResult
	ttl   time.Duration
}

type cachedResult struct {
	value     string
	timestamp time.Time
}

func NewCachingProxy(dbName string, ttl time.Duration) *CachingProxy {
	return &CachingProxy{
		db:    NewRealDatabase(dbName),
		cache: make(map[string]cachedResult),
		ttl:   ttl,
	}
}

func (p *CachingProxy) Query(sql string) string {
	if cached, ok := p.cache[sql]; ok {
		if time.Since(cached.timestamp) < p.ttl {
			fmt.Println("[Proxy] Cache HIT")
			return cached.value
		}
		fmt.Println("[Proxy] Cache EXPIRED")
		delete(p.cache, sql)
	}

	fmt.Println("[Proxy] Cache MISS - forwarding to real DB")
	result := p.db.Query(sql)
	p.cache[sql] = cachedResult{value: result, timestamp: time.Now()}
	return result
}

func (p *CachingProxy) Execute(sql string) bool {
	// Writes invalidate the cache
	p.cache = make(map[string]cachedResult)
	fmt.Println("[Proxy] Cache invalidated due to write")
	return p.db.Execute(sql)
}

func main() {
	var db Database = NewCachingProxy("production", 5*time.Second)

	// First query - cache miss
	r1 := db.Query("SELECT * FROM users")
	fmt.Println(r1)

	// Second query - cache hit
	r2 := db.Query("SELECT * FROM users")
	fmt.Println(r2)

	if r1 == r2 {
		fmt.Println("PASS: cached result matches")
	} else {
		panic("FAIL: cached result mismatch")
	}

	// Write invalidates cache
	db.Execute("INSERT INTO users VALUES (1, 'alice')")

	// Next query - cache miss again
	r3 := db.Query("SELECT * FROM users")
	fmt.Println(r3)
	fmt.Println("PASS: proxy pattern complete")
}
